aptos-move/framework/aptos-stdlib/sources/cryptography/curves.move

```diff
@@ -0,0 +1,217 @@
+    }
+
+    public fun scalar_from_u64<G>(value: u64): Scalar<G> {
+        Scalar<G> {
+            handle: scalar_from_u64_internal(value, get_group_id<G>())
+        }
+    }
+
+    public fun scalar_neg<T>(_scalar_1: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_neg_internal(get_group_id<T>())
+        }
+    }
+
+    public fun scalar_add<T>(_scalar_1: &Scalar<T>, _scalar_2: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_add_internal(_scalar_1.handle, _scalar_2.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_mul<T>(_scalar_1: &Scalar<T>, _scalar_2: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_mul_internal(_scalar_1.handle, _scalar_2.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_inv<T>(scalar: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_inv_internal(scalar.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_eq<T>(scalar_1: &Scalar<T>, scalar_2: &Scalar<T>): bool {
+        scalar_eq_internal(scalar_1.handle, scalar_2.handle, get_group_id<T>())
+    }
+
+    public fun bytes_into_scalar<T>(bytes: vector<u8>): Scalar<T> {
+        Scalar<T> {
+            handle: bytes_into_scalar_internal(bytes, get_group_id<T>())
+        }
+    }
+
+    // Point basics.
+    public fun point_identity<T>(): Point<T> {
+        Point<T> {
+            handle: point_identity_internal(get_group_id<T>())
+        }
+    }
+
+    public fun point_generator<T>(): Point<T> {
+        Point<T> {
+            handle: point_generator_internal(get_group_id<T>())
+        }
+    }
+
+    public fun point_add<T>(point_1: &Point<T>, point_2: &Point<T>): Point<T> {
+        Point<T> {
+            handle: point_add_internal(point_1.handle, point_2.handle, get_group_id<T>())
+        }
+    }
+
+    public fun point_mul<G>(_scalar: &Scalar<G>, _point: &Point<G>): Point<G> {
+        Point<G> {
+            handle: point_mul_internal(_scalar.handle, _point.handle, get_group_id<G>())
+        }
+    }
+
+    public fun scalar_to_bytes<T>(scalar: &Scalar<T>): vector<u8> {
+        scalar_to_bytes_internal(scalar.handle, get_group_id<T>())
+    }
+
+    public fun point_to_bytes<T>(point: &Point<T>): vector<u8> {
+        point_to_bytes_internal(point.handle, get_group_id<T>())
+    }
+
+    public fun bytes_into_point<T>(bytes: vector<u8>): Point<T> {
+        Point<T> {
+            handle: bytes_into_point_internal(bytes, get_group_id<T>())
+        }
+    }
+
+    public fun point_eq<T>(point_1: &Point<T>, point_2: &Point<T>): bool {
+        point_eq_internal(point_1.handle, point_2.handle, get_group_id<T>())
+    }
+
+    /// Group/bilinear mapping ID assignments.
+    /// The assignment here should match what is in `/aptos-move/framework/src/natives/cryptography/curves.rs`.
+    /// TODO: it is possible to retrieve move type info on rust end, so we do not need these ID assignments at all?
```

Comment:
Yes, it is possible: the native function needs to be a generic that takes type arguments. For example, `point_add_internal` would be `point_add_internal<T>`. Then, in Rust, there should be a way to get the struct's name. (I would have to look at my [std::debug::print PR here](https://github.com/move-language/move/pull/630/files#diff-5dbd856b718229a3b991331b8e7673d1cb1695300d6470f9896b7ca9dd068f0c), which does these sorts of things)

Reply:
I think that would also be faster than what you are doing now: which involves an extra `type_info` native call